McCorrections/src/LeptonWeights.cpp

```diff
@@ -60,51 +60,75 @@ MuonScaleFactorPOG::HistPtr MuonScaleFactorPOG::LoadWeight(const std::string& we
-        return electronSF.GetIdIsoSF(eventInfo.GetLeg(1).GetMomentum()) * tau_weight;
-    }
-    else if(channel == Channel::MuTau) {
-        double tau_weight = tauIdWeight->getSFvsPT(eventInfo.GetLeg(2).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return muonSF.GetIdIsoSF(eventInfo.GetLeg(1).GetMomentum()) * tau_weight;
-    }
-    else if(channel == Channel::TauTau) {
-        double tau_weight = tauIdWeight->getSFvsPT(eventInfo.GetLeg(1).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(1)->gen_match()))
-                          * tauIdWeight->getSFvsPT(eventInfo.GetLeg(2).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return tau_weight;
+    static const std::map<Period ,std::string>  period_label = { { Period::Run2016, "2016Legacy" },
+                                                     { Period::Run2017, "2017ReReco" },
+                                                     { Period::Run2018, "2018ReReco" } };
+     auto& provider = tau_sf_providers[discr][wp];
+     if(!provider) {
+         static const std::string s = "by";
+         const auto i = ToString(discr).find(s);
+         const std::string discr_name = ToString(discr).erase(i, s.length());
+         if(!period_label.count(period))
+             throw exception("LeptonWeights::GetTauIdProvider: Period %1% is not supported.") % period;
+         provider = std::make_shared<TauIDSFTool>(period_label.at(period), discr_name, ToString(wp), is_dm_binned, false);
+     }
+     return *provider;
+}
+
+double LeptonWeights::GetLegIdIsoWeight(LepCandidate leg, DiscriminatorWP VSe_wp, DiscriminatorWP VSmu_wp,
+                                        DiscriminatorWP VSjet_wp,
+                                        UncertaintySource unc_source, UncertaintyScale unc_scale)
+{
+    const UncertaintyScale current_scale = ApplyUncertaintyScale(leg->decayMode(), leg.GetMomentum().pt(),
+                                                                 leg.GetMomentum().eta(),leg->gen_match(), unc_source)
+                                           ? unc_scale : UncertaintyScale::Central;
+    if(leg->leg_type() == LegType::e)
+        return electronSF.GetIdIsoSF(leg.GetMomentum());
+
+    else if(leg->leg_type() == LegType::mu)
+        return  muonSF.GetIdIsoSF(leg.GetMomentum());
+
+    else if(leg->leg_type() == LegType::tau){
+        auto tauIdWeightVsEle = GetTauIdProvider(TauIdDiscriminator::byDeepTau2017v2p1VSe, VSe_wp);
+        auto tauIdWeightVsMu = GetTauIdProvider(TauIdDiscriminator::byDeepTau2017v2p1VSmu, VSmu_wp);
+        auto tauIdWeight = GetTauIdProvider(TauIdDiscriminator::byDeepTau2017v2p1VSjet, VSjet_wp);
+
+        double tau_id_weight_vs_ele = tauIdWeightVsEle.getSFvsEta(leg.GetMomentum().eta(), static_cast<int>(leg->gen_match()));
```

Comment:
- better to have control on gen_match also here. E.g. don't call `GetTauIdProvider(byDeepTau2017v2p1VSe, ...)` and then `getSF...` if gen_match does not correspond to the ones of tau.
- you should pass current_scale to the getSF method.